[JIT] Serialize ClassType as its qualname

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #29501 [WIP][JIT] Separate torchbind from Python
* **#30058 [JIT] Serialize ClassType as its qualname**

Differential Revision: [D18584269](https://our.internmc.facebook.com/intern/diff/D18584269)

Commit:
[JIT] Serialize ClassType as its qualname

[ghstack-poisoned]

Commit:
Update on "[WIP][JIT] Serialize ClassType as its qualname"


Differential Revision: [D18584269](https://our.internmc.facebook.com/intern/diff/D18584269)

[ghstack-poisoned]

Commit:
Update on "[WIP][JIT] Serialize ClassType as its qualname"


Differential Revision: [D18584269](https://our.internmc.facebook.com/intern/diff/D18584269)

[ghstack-poisoned]

Commit:
Update on "[WIP][JIT] Serialize ClassType as its qualname"


Differential Revision: [D18584269](https://our.internmc.facebook.com/intern/diff/D18584269)

[ghstack-poisoned]

zdevito commented on test/expect/TestScript.test_annot_ast_mypy_method.expect:
```diff
@@ -1,9 +1,9 @@
-foo(ClassType<FooModule> self, int[] x, (Tensor, Tensor) y) -> ((int[], int[]))
-foo(ClassType<FooModule> self, int? x, (Tensor, Tensor) y) -> ((int?, int?))
\ No newline at end of file
+foo( self, Tensor x, (Tensor, Tensor) y) -> ((Tensor, Tensor))
```
This looks wrong, the name is an empty string here?

jamesr66a replied:
It was too hard to figure out how to decay the name to the point where textual comparison would match in the presence of name mangling. Ideally we would get rid of these expect files altogether

zdevito commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -1393,7 +1393,7 @@ struct CAFFE2_API ClassType : public NamedType {
   std::string str() const override {
     const auto& n = name().value();
-    return std::string("ClassType<") + n.name() + ">";
+    return n.qualifiedName();
```
Can you call python_str() to keep them synced?

Commit:
Update on "[JIT] Serialize ClassType as its qualname"


Differential Revision: [D18584269](https://our.internmc.facebook.com/intern/diff/D18584269)

[ghstack-poisoned]
